db/keyval/filesystem/reader/reader.go

```diff
@@ -0,0 +1,189 @@
+					changed = append(changed, fData)
+					break
+				}
+			}
+		}
+		if !found {
+			removed = append(removed, FileEntry{key, value})
+		}
+	}
+	for _, fData := range f.Data {
+		var found bool
+		for key := range dataSet {
+			if fData.Key == key {
+				found = true
+				break
+			}
+		}
+		if !found {
+			changed = append(changed, fData)
+		}
+	}
+
+	return
+}
+
+// PathExists returns true if provided path exists, false otherwise
+func (r *Reader) PathExists(path string) bool {
+	if _, err := os.Stat(path); os.IsNotExist(err) {
+		return false
+	}
+	return true
+}
+
+// IsValid validates file before reading, whether it is not a temporary file
+func (r *Reader) IsValid(ev fsnotify.Event) (bool, error) {
+	// Silently skip empty event
+	if ev.Name == "" {
+		return false, nil
+	}
+	// Silently skip backup files ("~" at the end)
+	if strings.Compare(ev.Name[len(ev.Name)-1:], "~") == 0 {
+		return false, nil
+	}
+	// Validate file
+	path := strings.Split(ev.Name, "/")
+	if len(path) == 0 {
+		return false, fmt.Errorf("error validating file path: invalid input %s", ev.Name)
+	}
+	fileName := path[len(path)-1]
+	parts := strings.Split(fileName, ".")
+	if len(parts) == 0 {
+		return false, fmt.Errorf("error validating file in path %s: invalid format %s", ev.Name, fileName)
+	}
+	// Silently skip numeric files without extension (temporary files with PID of the editor)
+	if len(parts) == 1 && r.isNumeric(parts[0]) {
+		return false, nil
+	}
+	// Check file extension, skip
+	fileExtension := parts[len(parts)-1]
+	match, err := regexp.MatchString("sw.", fileExtension)
+	if err != nil {
+		return false, fmt.Errorf("error validating file %s: regex error: %v", ev.Name, err)
+	}
+	if match {
+		return false, nil
+	}
+	return true, nil
+}
+
+// ProcessFile reads file from filesystem and un-marshalls it to the required proto structure
+func (r *Reader) ProcessFile(path string) (File, error) {
+	dataSet := File{}
+	fileData, err := ioutil.ReadFile(path)
+	if err != nil {
+		return dataSet, fmt.Errorf("failed to read file from path %s: %v", path, err)
+	}
+
+	err = json.Unmarshal(fileData, &dataSet)
+	if err != nil {
+		return dataSet, fmt.Errorf("failed to unmarshal file %s: %v", path, err)
+	}
+
+	dataSet.Path = path
+	return dataSet, nil
+}
+
+// ProcessFilesInDir reads all file names from directory and processes them the ordinary way
+func (r *Reader) ProcessFilesInDir(path string) ([]File, error) {
+	fileInfoList, err := ioutil.ReadDir(path)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read files in directory %s", fileInfoList)
+	}
+	var files []File
+	for _, fileInfo := range fileInfoList {
+		file, err := r.ProcessFile(path + fileInfo.Name())
+		if err != nil {
+			return nil, err
+		}
+		files = append(files, file)
+	}
+
+	return files, nil
+}
+
+// IsDirectory checks if provided path is a file or a directory
+func (r *Reader) IsDirectory(path string) (bool, error) {
+	file, err := os.Open(path)
```

Comment:
This is not required to check if certain path is directory. Please use [os.Stat(path)](https://golang.org/pkg/os/#Stat).

Reply:
removed